Modules/readline.c

```diff
@@ -1019,10 +1019,16 @@ on_hook(PyObject *func)
 static int
 #if defined(_RL_FUNCTION_TYPEDEF)
 on_startup_hook(void)
+{
+#elif defined(__APPLE__) && defined(WITH_EDITLINE)
```

Comment:
Note that other distributors and users of Python on macOS may be linking to a libedit that is not the Apple-supplied libedit on macOS. For example, MacPorts supplies their own newer version of libedit. So tests like this might not be appropriate. It needs to be tested (see my comments on the issue).

Reply:
Have you established that there aren't also warnings when compiling with versions of editline other than the macOS system one? I literally don't have time to delve into this now as I'm traveling but I did try to do a quick compile with the current MacPorts version of editline I have installed with me and that also produced compile warnings though not exactly the same ones.

Reply:
* ./configure: Follow apple readline
* ./configure --with-readline=editline: Follow custom readline (macOS or latest BSD editline)
* ./configure --with-readline=readline: Follow custom readline (macOS or latest GNU readline)

Reply:
Or we can add a version check also: RL_READLINE_VERSION==0x0402, all macOS are using identical versions for built-in readline.

Reply:
cc @ronaldoussoren

Reply:
> I'm traveling but I did try to do a quick compile with the current MacPorts version of editline I have installed with me and that also produced compile warnings though not exactly the same ones.

Well, it's up to which commits of Python you are using.
After we reverted Irit's patch, there is no compiler warning for system one too. 
But it does not mean that the issue is solved because we don't use proper parameters for apple system built-in library and in the future clang compiler will not skip such a mistake.

Please see my comment: https://github.com/python/cpython/issues/108588#issuecomment-1697272676

Reply:
> Well, it's up to which commits of Python you are using.
> After we reverted Irit's patch, there is no compiler warning for system one too.

Umm, sorry, it was not reverted yet.

Reply:
> Have you established that there aren't also warnings when compiling with versions of editline other than the macOS system one?

I will test them and share soon.

Reply:
Okay with https://github.com/python/cpython/pull/108633/commits/b89dd4cafa72546069b67c6f47a98032fb1dca81
I detected the following warnings with homebrew libedit.

````
./Modules/readline.c:1269:21: warning: incompatible function pointer types assigning to 'rl_hook_func_t *' (aka 'int (*)(void)') from 'int (const char *, int)' [-Wincompatible-function-pointer-types]
    rl_startup_hook = on_startup_hook;
                    ^ ~~~~~~~~~~~~~~~
./Modules/readline.c:1271:23: warning: incompatible function pointer types assigning to 'rl_hook_func_t *' (aka 'int (*)(void)') from 'int (const char *, int)' [-Wincompatible-function-pointer-types]
    rl_pre_input_hook = on_pre_input_hook;
````

And with https://github.com/python/cpython/pull/108633/commits/67e3a54c99ece3eee7ba763a1dc7f0b1ec4848e2
there were no compiler warnings.

So I would like to propose updating the policy as I proposed from https://github.com/python/cpython/pull/108633#discussion_r1309106380
If people want to custom readline library passing the explicit option looks reasonable.